funsor/terms.py

```diff
@@ -30,15 +30,20 @@
 from funsor.six import getargspec, singledispatch
 
 
+def _cons_cache_key(*args):
+    return tuple(id(arg) if not isinstance(arg, Hashable) else arg for arg in args)
```

Comment:
I added this because `np.ndarray`s are not hashable, and should also work in case `Tensor.__hash__` is deprecated with https://github.com/pytorch/pytorch/issues/7733. Currently this would be a no-op for Tensor.